### PR TITLE
P9 scom refactor

### PR DESCRIPTION
**Pull request by lkammath**
1. pdbg submodule update to the latest upstream SHA
 2. p9 scom refactor to support multiple platforms at runtime
 3. Refactor initTargets() and other places referencing EDBG_DTB to move to PDBG_DTB

### CODE REVIEW

**Comment by lkammath** [comment]
@thejta @ojayanth please review

File: src/dll/edbgEcmdDll.C [hunk 1]
```diff
@@ -74,6 +74,9 @@ std::map<ecmdChipTarget, std::string, ecmdChipTargetCompare> eeproms;
 // Store if we have init'd the VPD info
 bool vpdInit = false;
 
+//chip type
+std::string chipType;
```
Comment (thejta) [inline src/dll/edbgEcmdDll.C:78]: Why a global variable with the chip type?

File: src/dll/edbgEcmdDll.C [hunk 2]
```diff
@@ -216,108 +162,59 @@ static uint32_t findChipUnitType(const ecmdChipTarget &i_target, uint64_t i_addr
...
+    uint32_t l_chip_id = 0;
+    uint32_t l_address = 0x100a; //Bits 12:19 tells chipID
+    ecmdDataBuffer l_data;
+    ecmdChipTarget l_target;
+
+    l_target.cageState =  l_target.nodeState = ECMD_TARGET_FIELD_VALID;
+    l_target.slotState = l_target.posState = ECMD_TARGET_FIELD_VALID;
+
+    //Read the 0x100a cfam to get the chip id
+    rc = dllGetCfamRegister(l_target, l_address, l_data);
```
Comment (thejta) [inline src/dll/edbgEcmdDll.C:184]: I don't know if cfam registers can be read from the host OS to do this determination.

File: src/dll/edbgEcmdDll.C [hunk 3]
```diff
@@ -216,108 +162,59 @@ static uint32_t findChipUnitType(const ecmdChipTarget &i_target, uint64_t i_addr
...
-  static int done = 0;
-
-  if (!done) {
-    done = 1;
-
-    // The user sets their device tree via this variable. If not set, fail
-    // Would be nice to also set via --device on the cmdline, but currently
-    // there is an order of operations problem.
-    // Longer term libpdbg will be able to auto-detect the correct thing to use.
-    char * devTree = getenv("EDBG_DTB");
```
Comment (thejta) [inline src/dll/edbgEcmdDll.C:291]: The EDBG_DTB variable can't just be removed.  There are places that set this variable that would break without it.  If pdbg_targets_init doesn't require the user to open the dtb any more, that's great - but PDBG_DTB would need to be set based on EDBG_DTB if it's set.

File: src/dll/edbgEcmdDll.C [hunk 4]
```diff
@@ -216,108 +162,59 @@ static uint32_t findChipUnitType(const ecmdChipTarget &i_target, uint64_t i_addr
...
 }
 
 // Load the device tree and initialise the targets
 static int initTargets(void) {
-  int fd;
-  void *fdt;
-  struct stat stat;
-  static int done = 0;
-
-  if (!done) {
```
Comment (thejta) [inline src/dll/edbgEcmdDll.C:284]: There were previously issues with calling pdbg_targets_init multiple times, so we had to manage that here.  Does pdbg now take care of that?

File: src/dll/edbgEcmdDll.C [hunk 5]
```diff
@@ -216,108 +162,59 @@ static uint32_t findChipUnitType(const ecmdChipTarget &i_target, uint64_t i_addr
...
-      return ECMD_FAILURE;
-    }
+  /*  Device tree can also be specified using PDBG_DTB environment variable
+   *  pointing to system device tree.  If system device tree is specified using
+   *  PDBG_DTB, then it will override the default device tree or the specified
+   *  device tree. NULL to use default which is used pdbg. */
+  pdbg_targets_init(NULL);
 
-    pdbg_targets_init(fdt);
+  uint32_t rc = getChipType(chipType);
```
Comment (thejta) [inline src/dll/edbgEcmdDll.C:213]: Why do we need to do this here?  Why isn't the chipType just checked when it's needed?

File: src/dll/edbgEcmdDll.C [hunk 6]
```diff
@@ -894,8 +791,8 @@ uint32_t queryConfigExistChipUnits(const ecmdChipTarget & i_target, struct pdbg_
       /* Skip targets with no ecmd equivalent */
       continue;
 
-    // If posState is set to VALID, check that our values match
-    // If posState is set to WILDCARD, we don't care
+    //If posState is set to VALID, check that our values match
```
Comment (thejta) [inline src/dll/edbgEcmdDll.C:794]: Spacing changes shouldn't be part of this PR.  Makes more look like it has changed than actually has.

File: src/dll/edbgEcmdDll.C [hunk 7]
```diff
@@ -1074,152 +970,50 @@ uint32_t dllCreateChipUnitScomAddress(const ecmdChipTarget & i_target, uint64_t
...
 
-  // Get the chip level pdbg target for the call to the pib write
-  if (fetchPdbgTarget(l_target, &target)) {
-    return out.error(EDBG_GENERAL_ERROR, FUNCNAME, "Unable to find PIB target\n");
-  }
-
-  // Make sure the pdbg target probe has been done and get the target state
-  if (pdbg_target_probe(target) != PDBG_TARGET_ENABLED) {
-    return out.error(ECMD_TARGET_NOT_CONFIGURED, FUNCNAME, "Target not configured!\n");
+  if (chipType == CHIP_TYPE_P9) {
```
Comment (thejta) [inline src/dll/edbgEcmdDll.C:1008]: Is there a way to get from pdbg the type of chip it is working with?

File: src/meson.build [hunk 8]
```diff
@@ -15,6 +14,7 @@ libekb = meson.get_compiler('cpp').find_library('ekb')
 
 libedbg = library(
             'edbg',
+            'dll/edbgEcmdDllScom_p9.C',
```
Comment (thejta) [inline src/meson.build:17]: Would it make more sense to put these files in the p9 directory?

File: src/dll/edbgEcmdDll.C (hunk 3 above)
Comment (lkammath) [reply src/dll/edbgEcmdDll.C:291]: Currently, pdbg itself will take care of using its own internal version of the device tree if not set explicitly.  so, EDBG doesn't have to keep its own version of the device tree copy. I have tested this on p9 system and it is working good

File: src/dll/edbgEcmdDll.C (hunk 4 above)
Comment (lkammath) [reply src/dll/edbgEcmdDll.C:284]: Is initTargets() called multiple times? i didn't see any issues after removing it. pdbg_targets_init() will give a warning if it is called second time but doesn't crash now.

File: src/dll/edbgEcmdDll.C (hunk 5 above)
Comment (lkammath) [reply src/dll/edbgEcmdDll.C:213]: I can do that. since init is called once in the program I thought it is the best place I can call and assign to the global variable.   I just tried to optimize the calling to the getChipType() every time and instead use the global var instead. But, i can change it if you think it is wrong approach

File: src/dll/edbgEcmdDll.C (hunk 6 above)
Comment (lkammath) [reply src/dll/edbgEcmdDll.C:794]: Will fix

File: src/dll/edbgEcmdDll.C (hunk 7 above)
Comment (lkammath) [reply src/dll/edbgEcmdDll.C:1008]: There is an API,  But all the changes are only pdbg downstream. since Alistair moved out of IBM not sure when it will get to upstream until unless someone manages upstream in place of Alistair

File: src/meson.build (hunk 8 above)
Comment (lkammath) [reply src/meson.build:17]: file name already has _p9.C. isn't that good?

File: src/dll/edbgEcmdDll.C (hunk 2 above)
Comment (lkammath) [reply src/dll/edbgEcmdDll.C:184]: Will fix it and use pdbg API's . There is an API, But all the changes are currently only in pdbg downstream. It will get to upstream in a couple of days (Delayed due to Aliaster moved out of IBM). I will use that API for now . it will break the native compilation now though but will get fixed once i update pdbg SHA to point to newer commit once the API is available

File: src/dll/edbgEcmdDll.C (hunk 1 above)
Comment (lkammath) [reply src/dll/edbgEcmdDll.C:78]: I can do that. since init is called once in the program I thought it is the best place I can call and assign to the global variable. I just tried to optimize the calling to the getChipType() every time and instead use the global var instead. But, i can change it if you think it is wrong approach.

Comment (thejta) [reply src/dll/edbgEcmdDll.C:78]: I would rather we get the information from pdbg when needed if it has an API for it.  At a minimum, it should be in a returned by a function so it is easy to change in the future if need.  Also,  it should be an enum value and not string we have to compare all the time

File: src/dll/edbgEcmdDll.C (hunk 2 above)
Comment (thejta) [reply src/dll/edbgEcmdDll.C]: If you make all this determination go through a common function it will be easy to swap in the pdbg API at a later date.

File: src/dll/edbgEcmdDll.C (hunk 3 above)
Comment (thejta) [reply src/dll/edbgEcmdDll.C:291]: You have missed the use case where EDBG_DTB=none.  That is used by the code to not even do the pdbg init because not all features of the code use pdbg, like VPD update.  If you want to consolidate down to just PDBG_DTB, that is fine - but you need to update the other places in the repo that set EDBG_DTB and preserve the none use case.

File: src/dll/edbgEcmdDll.C (hunk 4 above)
Comment (thejta) [reply src/dll/edbgEcmdDll.C:284]: If calling it multiple times is a valid code path, we don't want it throwing a warning at the user for something that is valid.  It looks like now initTargets is only called on dll load, it should be okay.  But if you are seeing the warning in your testing, that need to be fixed.

File: src/dll/edbgEcmdDll.C (hunk 5 above)
Comment (thejta) [reply src/dll/edbgEcmdDll.C:197]: Please have a function that checks the product generation (P9, P10, etc..) that is just called when needed.  That can stay consistent in the code and then call the correct underlying hardware or pdbg function to make that determination.

File: src/dll/edbgEcmdDll.C (hunk 7 above)
Comment (thejta) [reply src/dll/edbgEcmdDll.C:992]: This goes back to have a common function to get this information and have it return an actual enum value as well, instead of a string we are comparing against.

File: src/meson.build (hunk 8 above)
Comment (thejta) [reply src/meson.build:17]: Why have a p9, etc.. directory if the p9, etc.. specific code isn't in it?

File: src/dll/edbgEcmdDll.C (hunk 1 above)
Comment (lkammath) [reply src/dll/edbgEcmdDll.C:78]: Done. I am using pdbg api only if you had checked the code. just that we need to get api to upstream and same will be done

File: src/dll/edbgEcmdDll.C (hunk 2 above)
Comment (lkammath) [reply src/dll/edbgEcmdDll.C]: Done

File: src/dll/edbgEcmdDll.C (hunk 4 above)
Comment (lkammath) [reply src/dll/edbgEcmdDll.C:284]: atleast running HW access commands like getcfam, get/putscom, get/putmem, cipinstrucstart/stop all I did not see any issue nor has thrown any warning

File: src/dll/edbgEcmdDll.C (hunk 5 above)
Comment (lkammath) [reply src/dll/edbgEcmdDll.C:197]: done

File: src/meson.build (hunk 8 above)
Comment (lkammath) [reply src/meson.build:17]: Done. move to p9 folder

File: src/dll/edbgEcmdDll.C (hunk 3 above)
Comment (lkammath) [reply src/dll/edbgEcmdDll.C:291]: I did not see any reference to EDBG_DTB anywhere else in the code. If you can let me know were and how to test vpd i can verify. If you want to keep both the codes.  If the VPD functions are not using the devtree or the current pdbg_targets_init() function then i would feel the current change is safe

Comment (thejta) [reply src/dll/edbgEcmdDll.C:291]: If you grep the repo EDBG_DTB is set in command scripts and the like that call into edbg.  If you preserve the 'none' functionality, that should be fine - it doesn't have to be directly tested.  That would be preserved preferably through replacing the EDBG_DTB reference with PDBG_DTB and then preserving a PDBG_DTB=none code path that doesn't call pdbg init.

Comment (lkammath) [reply src/dll/edbgEcmdDll.C:291]: Yes, Thats preserved

Comment (lkammath) [reply src/dll/edbgEcmdDll.C:291]: Just realized. we didn't touch the original edbgWrapper.sh and we kept as it is and all the modifications done w.r.t to bmc is in  edbgBmcWrapper.sh

Comment (lkammath) [reply src/dll/edbgEcmdDll.C:291]: root@tacoma1z-bigcorebmc1:/usr/bin# export EDBG_DTB=none
root@tacoma1z-bigcorebmc1:/usr/bin# export EDBG_HOME=/usr/
root@tacoma1z-bigcorebmc1:/usr/bin# export ECMD_EXE=/usr/bin/edbg 

root@tacoma1z-bigcorebmc1:/usr/bin# env
SHELL=/bin/sh
EDITOR=vi
PWD=/usr/bin
LOGNAME=root
HOME=/home/root
ECMD_EXE=/usr/bin/edbg
SSH_CONNECTION=::ffff:9.3.62.216 32924 ::ffff:9.3.86.184 22
TERM=xterm
USER=root
SHLVL=1
SYSTEMD_PAGER=
EDBG_HOME=/usr/
PS1=\u@\h:\w\$ 
SSH_CLIENT=::ffff:9.3.62.216 32924 22
PATH=/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin
EDBG_DTB=none
SSH_TTY=/dev/pts/3
OLDPWD=/usr/bin/bkup
_=/usr/bin/env

root@tacoma1z-bigcorebmc1:/usr/bin# getscom pu 50009 -all
pu      k0:n0:s0:p00       0x8540508300000000
pu      k0:n0:s0:p01       0x8436020300000000
/usr/bin/edbg getscom pu 50009 -all
